src/python/py/models/quantized_model.py

```diff
@@ -52,51 +52,86 @@ def __init__(self):
+        self.gate_proj = QuantizedTensorModule(bits, group_size) if not is_lora else QuantizedLoRAModule(bits, group_size)
+        self.up_proj = QuantizedTensorModule(bits, group_size) if not is_lora else QuantizedLoRAModule(bits, group_size)
+        self.down_proj = QuantizedTensorModule(bits, group_size) if not is_lora else QuantizedLoRAModule(bits, group_size)
+        self.fc1 = QuantizedTensorModule(bits, group_size) if not is_lora else QuantizedLoRAModule(bits, group_size)
+        self.fc2 = QuantizedTensorModule(bits, group_size) if not is_lora else QuantizedLoRAModule(bits, group_size)
 
 
 class QuantizedDecoderLayer:
-    def __init__(self, layer_id, bits, group_size):
+    def __init__(self, layer_id, bits, group_size, is_lora):
         self.layer_id = layer_id
         self.input_layernorm = TensorModule()
-        self.self_attn = QuantizedAttention(bits, group_size)
+        self.self_attn = QuantizedAttention(bits, group_size, is_lora)
         self.post_attention_layernorm = TensorModule()
-        self.mlp = QuantizedMLP(bits, group_size)
+        self.mlp = QuantizedMLP(bits, group_size, is_lora)
 
     def is_empty(self):
         return self.input_layernorm.weight is None
 
 
 class QuantizedModel:
-    def __init__(self, quant_type, input_path, bits, group_size, q_size, kv_size, intermediate_size, num_layers):
+    def __init__(self, quant_type, input_path, bits, group_size, q_size, kv_size, intermediate_size, num_layers, adapter_path):
         self.quant_type = quant_type
         self.embedding = TensorModule()
         self.final_norm = TensorModule()
         self.lm_head = TensorModule()
         self.layers = {}
         self.num_layers = num_layers
 
+        is_lora = adapter_path is not None
+
+        self.map_to_modules(quant_type, input_path, bits, group_size, q_size, kv_size, intermediate_size, adapter_path)
+
+        if adapter_path is not None:
+            self.map_to_modules(quant_type, adapter_path, bits, group_size, q_size, kv_size, intermediate_size, adapter_path)
+
+        # Set LM head weights + biases if not already set
+        if isinstance(self.lm_head, TensorModule) and self.lm_head.weight is None:
+            # Embedding and LM head share same weights + biases (lm_head.weight == embedding.weight and lm_head.bias == embedding.bias)
+            self.lm_head.weight = self.embedding.weight
+            if self.lm_head.bias is not None:
+                self.lm_head.bias = self.embedding.bias
+
+        # Sort list of layers by layer id
+        self.layers = list(self.layers.values())
+        self.layers.sort(key=lambda m: m.layer_id)
+
+        # Set properties of each layer based on quantization type
+        self.set_properties(is_lora)
+
+    def map_to_modules(self, quant_type, file_path, bits, group_size, q_size, kv_size, intermediate_size, adapter_path):
         layer_id = 0
-        for weight_file in os.listdir(input_path):
+        is_lora = adapter_path is not None
```

Comment:
If `self.is_lora` is used, then this can be removed and any references to `is_lora` can be replaced with `self.is_lora`.